valhalla/common/telescope_states.py

```diff
@@ -15,116 +15,188 @@
+                         _source=['timestamp', 'telescope', 'enclosure', 'site', 'type', 'reason'],
+                         sort=['site', 'enclosure', 'telescope', 'timestamp'])  # noqa
+        event_data.extend(data['hits']['hits'])
+        total_events = data['hits']['total']
+        events_read = min(query_size, total_events)
+        scroll_id = data.get('_scroll_id', 0)
+        while events_read < total_events:
+            data = es.scroll(scroll_id=scroll_id, scroll='1m') # noqa
+            scroll_id = data.get('_scroll_id', 0)
+            event_data.extend(data['hits']['hits'])
+            events_read += len(data['hits']['hits'])
+        return event_data
+
+    def _belongs_in_lump(self, event_source, lump_data, dt=60):
+        if str(lump_data['telescope']) != str(self._telescope(event_source)):
+            return False
+
+        time_diff = (self._es_to_datetime(event_source['timestamp']) - lump_data['latest_timestamp']).total_seconds()
+
+        # If the event is close enough to the latest timestamp in the lump, it belongs in that lump.
+        if time_diff < dt:
+            return True
+
+        # If the event is far in time from the lump, but has the same reason, that means it came from the same
+        # scheduling run as that timestamp and belongs in the lump.
+        if time_diff >= dt and event_source['reason'] in lump_data['reasons']:
+            return True
+
+        # If the event is an ENCLOSURE_INTERLOCK, and the lump includes a SEQUENCER_UNAVAILABLE, it is part of the
+        # lump. This is because the two states are redundant.
+        if event_source['type'] == 'ENCLOSURE_INTERLOCK' and 'SEQUENCER_UNAVAILABLE' in lump_data['types']:
+            return True
+
+        return False
+
+    def _categorize(self, lump):
+        # TODO: Categorize each lump in a useful way for network users.
+        event_type, event_reason = lump['types'][0], lump['reasons'][0]
+        for this_type, this_reason in zip(lump['types'], lump['reasons']):
+            # If the state in ENCLOSURE INTERLOCK, wait to instead categorize as SEQUENCER_UNAVAILABLE.
+            if this_type.upper() != 'ENCLOSURE_INTERLOCK':
+                event_type, event_reason = this_type, this_reason
+                break
+        return event_type, event_reason
+
+    def _lump_end(self, lump, next_event=None):
+        if not next_event or str(lump['telescope']) != str(self._telescope(next_event)):
+            return self.end
+        return self._es_to_datetime(next_event['timestamp'])
+
+    def _set_lump(self, event):
+        return {
+            'reasons': [event['_source']['reason']],
+            'types': [event['_source']['type']],
+            'start': self._es_to_datetime(event['_source']['timestamp']),
+            'telescope': self._telescope(event['_source']),
+            'latest_timestamp': self._es_to_datetime(event['_source']['timestamp'])
+        }
+
+    def _update_lump(self, lump, event):
```

Comment:
seems like a lot of __lump methods are not really class methods (don't need the self reference). Maybe it would be better to make a Lump class with its own instance methods to update, get data from, or pull some methods out of the class, or make them static methods.